src/agent/srcview/src/pdbcache.rs

```diff
@@ -5,30 +5,92 @@ use std::collections::BTreeMap;
 use std::fs::File;
 use std::path::{Path, PathBuf};
 
-use anyhow::{format_err, Result};
+use anyhow::{bail, format_err, Result};
 use log::*;
 use pdb::{FallibleIterator, SymbolData, PDB};
 use serde::{Deserialize, Serialize};
 
 use crate::SrcLine;
 
+// NOTE: We're using strings as the keys for now while we build the trees, since
+// PathBuf comparisons are expensive.
+#[derive(Default)]
+struct PdbCacheBuilder {
+    offset_to_line: BTreeMap<usize, Vec<SrcLine>>,
+    symbol_to_lines: BTreeMap<String, Vec<SrcLine>>,
+    path_to_symbols: BTreeMap<String, Vec<String>>,
+    path_to_lines: BTreeMap<String, Vec<usize>>,
+}
+
+impl PdbCacheBuilder {
+    fn build(self) -> PdbCache {
+        PdbCache {
+            offset_to_line: self.offset_to_line,
+            symbol_to_lines: self.symbol_to_lines,
+            path_to_symbols: self
+                .path_to_symbols
+                .into_iter()
+                .map(|(p, s)| (PathBuf::from(p), s))
+                .collect(),
+            path_to_lines: self
+                .path_to_lines
+                .into_iter()
+                .map(|(p, l)| (PathBuf::from(p), l))
+                .collect(),
+        }
+    }
+
+    fn update_from_iter<I: FallibleIterator<Item = pdb::LineInfo, Error = pdb::Error>>(
+        &mut self,
+        program: &pdb::LineProgram,
+        string_table: &pdb::StringTable,
+        address_map: &pdb::AddressMap,
+        proc_name: &str,
+        mut lines: I,
+    ) -> Result<()> {
+        let symbol_to_lines = self
+            .symbol_to_lines
+            .entry(proc_name.to_string())
+            .or_default();
+
+        while let Some(line_info) = lines.next()? {
+            let rva = line_info
+                .offset
+                .to_rva(address_map)
+                .ok_or_else(|| format_err!("invalid RVA: {:?}", line_info))?;
+            let file_info = program.get_file_info(line_info.file_index)?;
+            let file_name = file_info.name.to_string_lossy(string_table)?;
+
+            let path = file_name.into_owned();
+            let offset_to_line = self.offset_to_line.entry(rva.0 as usize).or_default();
+            let path_to_symbols = self.path_to_symbols.entry(path.clone()).or_default();
+            let path_to_lines = self.path_to_lines.entry(path.clone()).or_default();
+
+            for line in line_info.line_start as usize..line_info.line_end as usize + 1 {
+                let srcloc = SrcLine::new(path.clone(), line);
+
+                offset_to_line.push(srcloc.clone());
+                symbol_to_lines.push(srcloc.clone());
+                path_to_symbols.push(proc_name.to_string());
```

Comment:
```suggestion
                path_to_symbols.push(proc_name.to_string());
```

This line in particular could be hoisted out of the loop. It might be worth making `path_to_symbols` map to a `BTreeSet`?